controllers/tests/searchProductController.test.js

```diff
@@ -0,0 +1,185 @@
+      req = { params: { keyword: "a" } }; // minimum input of 1 character
+      productModel.find = jest
+        .fn()
+        .mockReturnValue({ select: jest.fn().mockResolvedValue(mockProducts) });
+
+      await searchProductController(req, res);
+
+      // Assertions
+      expect(res.status).toBeCalledWith(200);
+      expect(res.json).toBeCalledWith({
+        success: true,
+        results: mockProducts,
+      });
+    });
+
+    it("Should fetch associated products given maximum valid input that matches product name or description", async () => {
+      req = { params: { keyword: "a".repeat(100) } }; // maximum input of 100 characters
+      productModel.find = jest.fn().mockReturnValue({
+        select: jest.fn().mockResolvedValue([mockProducts[2]]),
+      });
+
+      await searchProductController(req, res);
+
+      // Assertions
+      expect(res.status).toBeCalledWith(200);
+      expect(res.json).toBeCalledWith({
+        success: true,
+        results: [mockProducts[2]],
+      });
+    });
+
+    it("Should return 400 error given just above maximum valid input", async () => {
+      req = { params: { keyword: "a".repeat(101) } }; // just above maximum input of 100 characters
+
+      await searchProductController(req, res);
+
+      // Assertions
+      expect(res.status).toBeCalledWith(400);
+      expect(res.json).toBeCalledWith({
+        success: false,
+        message: "Keyword is too long",
+      });
+    });
+  });
+
+  describe("Functional test cases", () => {
+    it("Should deal with input appropriately that has special characters", async () => {
+      req = { params: { keyword: "(c+)+" } }; // input with special characters
+      productModel.find = jest.fn().mockReturnValue({
+        select: jest.fn().mockResolvedValue([]),
+      });
+
+      await searchProductController(req, res);
+
+      // Assertions
+      expect(productModel.find).toBeCalledWith({
+        $or: [
+          { name: { $regex: "\\(c\\+\\)\\+", $options: "i" } },
+          { description: { $regex: "\\(c\\+\\)\\+", $options: "i" } },
+        ],
+      });
+      expect(res.status).toBeCalledWith(200);
+      expect(res.json).toBeCalledWith({
+        success: true,
+        results: [],
+      });
+    });
+
+    it("Should fetch associated products given valid input that matches product name or description, regardless of case", async () => {
+      req = { params: { keyword: "giRaFfe" } }; // input with special characters
+      productModel.find = jest.fn().mockReturnValue({
+        select: jest.fn().mockResolvedValue([mockProducts[0]]),
+      });
+
+      await searchProductController(req, res);
+
+      // Assertions
+      expect(res.status).toBeCalledWith(200);
+      expect(res.json).toBeCalledWith({
+        success: true,
+        results: [mockProducts[0]],
+      });
+    });
+
+    it("Should return error response when there is error fetching associated products", async () => {
```

Comment:
small good practice thing but maybe we can leave the error message here to be relevant such as "Error fetching product" in the mock and assert the same error message?